src/main/java/com/faforever/client/replay/ReplayService.java

```diff
@@ -373,6 +400,57 @@ private void runFafReplayFile(Path path) throws IOException, CompressorException
     gameService.runWithReplay(tempSupComReplayFile, replayId, gameType, version, modVersions, simMods, mapName);
   }
 
+  private void runHostFromOnlineReplay(int replayId) {
+    downloadReplay(replayId)
+        .thenAccept((path) -> {
+          try {
+            runHostFromReplayFile(path);
+          } catch (IOException | CompressorException e) {
+            throw new CompletionException(e);
+          }
+        })
+        .exceptionally(throwable -> {
+          if (throwable.getCause() instanceof FileNotFoundException) {
+            log.warn("Replay not available on server yet", throwable);
+            notificationService.addImmediateWarnNotification("replayNotAvailable", replayId);
+          } else {
+            log.error("Replay could not be started", throwable);
+            notificationService.addImmediateErrorNotification(throwable, "replayCouldNotBeStarted", replayId);
+          }
+          return null;
+        });
+  }
+
+  private void runHostFromFafReplayFile(Path path) throws IOException, CompressorException {
+    ReplayDataParser replayData = replayFileReader.parseReplay(path);
+
+    ReplayMetadata replayMetadata = replayData.getMetadata();
+    String gameType = replayMetadata.getFeaturedMod();
+    String mapName = parseMapFolderName(replayData);
+    Set<String> simMods = parseModUIDs(replayData);
+
+    modService.enableSimMods(simMods);
+    modService.getFeaturedMod(gameType)
+        .thenAccept(featuredModBean -> gameService.hostGame(new NewGameInfo(replayMetadata.getTitle(),
+            null, featuredModBean, mapName, simMods, GameVisibility.PUBLIC,null,
+            null,false)));
+  }
```

Comment:
This would be better done to return a completable future and chain the gameService.hostGame together so that it can properly wait until the game is hosted if necessary.
```suggestion
  private CompletableFuture<Void> runHostFromFafReplayFile(Path path) throws IOException, CompressorException {
    ReplayDataParser replayData = replayFileReader.parseReplay(path);

    ReplayMetadata replayMetadata = replayData.getMetadata();
    String gameType = replayMetadata.getFeaturedMod();
    String mapName = parseMapFolderName(replayData);
    Set<String> simMods = parseModUIDs(replayData);

    modService.enableSimMods(simMods);
    return modService.getFeaturedMod(gameType)
        .thenCompose(featuredModBean -> {
        NewGameInfo newGameInfo = new NewGameInfo(
            replayMetadata.getTitle(),
            null, featuredModBean, 
            mapName, 
            simMods, 
            GameVisibility.PUBLIC,
            null,
            null,
            false
        );
        return gameService.hostGame(newGameInfo);
    });
  }
```

Reply:
Note I wrote this in github so would need to double check syntax is correct